cmd/kops/create_cluster.go

```diff
@@ -283,87 +286,102 @@ func RunCreateCluster(f *util.Factory, out io.Writer, c *CreateClusterOptions) e
 	var nodes []*api.InstanceGroup
 	var instanceGroups []*api.InstanceGroup
 
-	if len(c.MasterZones) == 0 {
-		if len(masters) == 0 {
-			// We default to single-master (not HA), unless the user explicitly specifies it
-			// HA master is a little slower, not as well tested yet, and requires more resources
-			// Probably best not to make it the silent default!
-			for _, subnet := range cluster.Spec.Subnets {
-				g := &api.InstanceGroup{}
-				g.Spec.Role = api.InstanceGroupRoleMaster
-				g.Spec.Subnets = []string{subnet.Name}
-				g.Spec.MinSize = fi.Int32(1)
-				g.Spec.MaxSize = fi.Int32(1)
-				g.ObjectMeta.Name = "master-" + subnet.Name // Subsequent masters (if we support that) could be <zone>-1, <zone>-2
-				instanceGroups = append(instanceGroups, g)
-				masters = append(masters, g)
-
-				// Don't force HA master
-				break
-			}
-		}
-	} else {
-		if len(masters) == 0 {
-			// Use the specified master zones (this is how the user gets HA master)
+	// Build the master subnets
+	// The master zones is the default set of zones unless explicitly set
+	// The master count is the number of master zones unless explicitly set
+	// We then round-robin around the zones
+	if len(masters) == 0 {
+		var masterSubnets []*api.ClusterSubnetSpec
+		if len(c.MasterZones) != 0 {
 			for _, subnetName := range c.MasterZones {
-				g := &api.InstanceGroup{}
-				g.Spec.Role = api.InstanceGroupRoleMaster
-				g.Spec.Subnets = []string{subnetName}
-				g.Spec.MinSize = fi.Int32(1)
-				g.Spec.MaxSize = fi.Int32(1)
-				g.ObjectMeta.Name = "master-" + subnetName
-				instanceGroups = append(instanceGroups, g)
-				masters = append(masters, g)
+				subnet := findSubnet(cluster, subnetName)
+				if subnet == nil {
+					// Should have been caught already
+					return fmt.Errorf("master-zone %q not included in zones", subnetName)
+				}
+				masterSubnets = append(masterSubnets, subnet)
+			}
+
+			if c.MasterCount != 0 && c.MasterCount < int32(len(masterSubnets)) {
+				return fmt.Errorf("specified %d master zones, but also requested %d masters.  If specifying both, the count should match.", len(masterSubnets), c.MasterCount)
```

Comment:
Nothing to do with quorum.

It doesn't make any sense to specify mastercount=2 and 3 master zones explicitly.  2 master zones implies 2 masters.  The third master zone would be ignored.  It just looks like human error.